### ISSUE TITLE
Figure out why some text field entries are in DB twice

### ISSUE THREAD

**Issue by davidskalinder** [davidskalinder/mpeds-coder#85]
With same timestamps, even.

@olderwoman sez:
> Hi, David, we have identified a problem with the export file that probably points to a deeper problem. The overt problem is that some start and end date fields have multiple dates in them. Most of the time these are the same date twice; occasionally the same date three times, occasionally different dates. When I go back and look at MAI for what the coder did, there is only one date showing on the MAI interface. It seems likely that this problem is coming from multiple coder lines for the same variable (start or end date) and event, which would be generated when the coder returns to edit an event, which they do sometimes to correct their work or in response to supervision. Our guess is that failure to include timestamp is giving us the problem that we don’t know when a particular line was written and this also probably points to the likelihood that there are other variables that may have this problem. We need to know whether/how coder lines are overwritten or appended if a coder comes back and edits their work. And, of course, the order in which the coding happened. 

See email with attachments for some examples.

**Comment by olderwoman** [comment]
Here is the updated file with problematic dates including timestamps.
[bpp_mai_2020-05-27_164412bad_dates.xlsx](https://github.com/davidskalinder/mpeds-coder/files/4699246/bpp_mai_2020-05-27_164412bad_dates.xlsx)

**Comment by olderwoman** [comment]
and here is the file that includes all the lines for any coder/article with a bad date
[bpp_mai_2020-05-27_164412bad_dates_allarts.xlsx](https://github.com/davidskalinder/mpeds-coder/files/4699256/bpp_mai_2020-05-27_164412bad_dates_allarts.xlsx)

**Comment by davidskalinder** [comment]
> So a look at the "long file" coder-table_2020-04-095615.csv makes it look like this is due to multiple records with start-date or end-date. This can happen when a coder changes their mind. So we do seem to have the problem of not being able to distinguish appending multiple responses from replacing responses, so far as I can tell. I think the long file will be instructive in figuring things out. For dates we probably want to overwrite and take the one with last timestamp. But is a record generated when a coder deletes marked text in the text-marking fields?

I think this note (from #87) is for this issue?  But helpful, so I'm moving it here.  I'll have a closer look when I crack it open next week.

**Comment by davidskalinder** [comment]
Looping in @alexhanna.

Some back-end diagnostics for this:

```
/* Count of rows with multiple-valued variables */
SELECT a.variable, count(*)
  FROM coder_event_creator AS a
  INNER JOIN
    (SELECT coder_id
       , event_id
       , variable
       , COUNT(id) AS n
       FROM coder_event_creator
       GROUP BY coder_id
       , event_id
       , variable
       HAVING n > 1
     ) AS b
  ON a.coder_id = b.coder_id
  AND a.event_id = b.event_id
  AND a.variable = b.variable
  GROUP BY a.variable;
```

On our production DB atm this produces:

```
+----------------------+----------+
| variable             | count(*) |
+----------------------+----------+
| actions-text         |     3130 |
| actor-text           |     5808 |
| bystander-text       |      103 |
| end-date             |       33 |
| end-date-estimated   |        2 |
| issue-text           |     2225 |
| location-text        |      970 |
| named-org-text       |     2511 |
| police-actions-text  |      401 |
| size-text            |      355 |
| start-date           |       99 |
| start-date-estimated |        4 |
| target-text          |     1073 |
| time-text            |      312 |
+----------------------+----------+
```

The ones ending in `-text` are text-capture variables for which we expect multiple values.  So the (guaranteed) problematic rows are found by this query:

```
/* Rows with non-text-capture multi-entry variables */
SELECT *
  FROM coder_event_creator AS a
  INNER JOIN
    (SELECT coder_id
       , event_id
       , variable
       , COUNT(id) AS n
       FROM coder_event_creator
       WHERE RIGHT(variable, 5) != "-text"
       GROUP BY coder_id
       , event_id
       , variable
       HAVING n > 1
     ) AS b
  ON a.coder_id = b.coder_id
  AND a.event_id = b.event_id
  AND a.variable = b.variable;
```

This gives us 138 rows atm, of which @olderwoman's sheet above is a strict subset (as it should be).  So that defines the scope of the confirmable problem.

I had some careless pokes at the development interface and I can't reproduce this yet, neither on the text-select items (comfortingly) nor on the start/end date fields (less comfortingly).  I will poke more carefully.

**Comment by davidskalinder** [comment]
More diagnostics:

```
/* Dodgy rows that don't share a timestamp with their twin */
SELECT a2.*
  FROM coder_event_creator AS a2
  INNER JOIN (
    /* Get singleton timestamps */
    SELECT a.coder_id, a.event_id, a.variable, a.timestamp
      FROM coder_event_creator AS a
      INNER JOIN (
        /* Count multi-entry non-text-select vars */
        SELECT coder_id
          , event_id
          , variable
          , COUNT(id) AS n
          FROM coder_event_creator
          WHERE RIGHT(variable, 5) != "-text"
          GROUP BY coder_id
          , event_id
          , variable
          HAVING n > 1
        ) AS b
      ON a.coder_id = b.coder_id
      AND a.event_id = b.event_id
      AND a.variable = b.variable
      GROUP BY a.coder_id, a.event_id, a.variable, a.timestamp
      HAVING COUNT(a.id) = 1
    ) AS b2
  ON a2.coder_id = b2.coder_id
  AND a2.event_id = b2.event_id
  AND a2.variable = b2.variable;
```

This yields:
```
+-------+------------+----------+----------------------+------------+------+----------+---------------------+
| id    | article_id | event_id | variable             | value      | text | coder_id | timestamp           |
+-------+------------+----------+----------------------+------------+------+----------+---------------------+
|  9952 |       1717 |      595 | start-date           | 1995-03-15 | NULL |        9 | 2020-02-25 02:07:14 |
|  9955 |       1717 |      595 | start-date           | 1995-03-15 | NULL |        9 | 2020-02-25 02:07:15 |
|  9960 |       1717 |      595 | start-date           | 1995-03-09 | NULL |        9 | 2020-02-25 02:07:15 |
| 74312 |       2764 |     4838 | end-date-estimated   | yes        | NULL |        9 | 2020-05-21 21:48:16 |
| 74314 |       2764 |     4838 | end-date-estimated   | yes        | NULL |        9 | 2020-05-21 21:48:19 |
| 74312 |       2764 |     4838 | end-date-estimated   | yes        | NULL |        9 | 2020-05-21 21:48:16 |
| 74314 |       2764 |     4838 | end-date-estimated   | yes        | NULL |        9 | 2020-05-21 21:48:19 |
| 78285 |       2755 |     5100 | start-date-estimated | yes        | NULL |        9 | 2020-05-26 10:46:04 |
| 78291 |       2755 |     5100 | start-date-estimated | yes        | NULL |        9 | 2020-05-26 10:46:36 |
| 78285 |       2755 |     5100 | start-date-estimated | yes        | NULL |        9 | 2020-05-26 10:46:04 |
| 78291 |       2755 |     5100 | start-date-estimated | yes        | NULL |        9 | 2020-05-26 10:46:36 |
| 42172 |       4280 |     2720 | start-date-estimated | yes        | NULL |       17 | 2020-04-01 04:24:11 |
| 42175 |       4280 |     2720 | start-date-estimated | yes        | NULL |       17 | 2020-04-01 04:24:19 |
| 42172 |       4280 |     2720 | start-date-estimated | yes        | NULL |       17 | 2020-04-01 04:24:11 |
| 42175 |       4280 |     2720 | start-date-estimated | yes        | NULL |       17 | 2020-04-01 04:24:19 |
+-------+------------+----------+----------------------+------------+------+----------+---------------------+
```

So all the twin entries happen within a second of each other except for the handful of checkbox entries.

**Comment by davidskalinder** [comment]
And another check:

```
/* Dodgy rows that don't share a value with their twin */
SELECT DISTINCT a2.*
  FROM coder_event_creator AS a2
  INNER JOIN (
    /* Get singleton values */
    SELECT a.coder_id, a.event_id, a.variable, a.value
      FROM coder_event_creator AS a
      INNER JOIN (
        /* Count multi-entry non-text-select vars */
        SELECT coder_id
          , event_id
          , variable
          , COUNT(id) AS n
          FROM coder_event_creator
          WHERE RIGHT(variable, 5) != "-text"
          GROUP BY coder_id
          , event_id
          , variable
          HAVING n > 1
        ) AS b
      ON a.coder_id = b.coder_id
      AND a.event_id = b.event_id
      AND a.variable = b.variable
      GROUP BY a.coder_id, a.event_id, a.variable, a.value
      HAVING COUNT(a.id) = 1
    ) AS b2
  ON a2.coder_id = b2.coder_id
  AND a2.event_id = b2.event_id
  AND a2.variable = b2.variable;
```

Which gets us:

```
+------+------------+----------+------------+------------+------+----------+---------------------+
| id   | article_id | event_id | variable   | value      | text | coder_id | timestamp           |
+------+------------+----------+------------+------------+------+----------+---------------------+
| 9956 |       1717 |      595 | end-date   | 1995-03-22 | NULL |        9 | 2020-02-25 02:07:15 |
| 9958 |       1717 |      595 | end-date   | 1995-03-15 | NULL |        9 | 2020-02-25 02:07:15 |
| 9959 |       1717 |      595 | end-date   |            | NULL |        9 | 2020-02-25 02:07:15 |
| 9952 |       1717 |      595 | start-date | 1995-03-15 | NULL |        9 | 2020-02-25 02:07:14 |
| 9955 |       1717 |      595 | start-date | 1995-03-15 | NULL |        9 | 2020-02-25 02:07:15 |
| 9960 |       1717 |      595 | start-date | 1995-03-09 | NULL |        9 | 2020-02-25 02:07:15 |
+------+------------+----------+------------+------------+------+----------+---------------------+
6 rows in set (0.08 sec)
```

So there's only one event with any dodgy values that are different from each other.  That at least means that cleaning it up will be easy, assuming the problem doesn't extend beyond these fields...

**Comment by davidskalinder** [comment]
@alexhanna, I see that the campus protest database also has a lot of these (you guys have a lot more of what I assume are intentionally multi-valued variables, but I assume `start-date` is not one of those).  There are 390 duplicate `start-date` entries, and they are all straight-up duplicates: same value, same timestamp, same everything but the primary key.  You might want to run the first query above and eyeball the fields to check other things that you know shouldn't be multi-entry.  (The thing that you'd need to change to make the other queries work is that `WHERE` condition, since the current one only gets rid of text-select variables.)

I'm a bit stumped about why I can't reproduce this, why it would happen only sporadically, and why it would usually happen with the same value in the same second but occasionally with longer gaps.  I wonder if some `DELETE` statements might be getting skipped due to connection trouble or something?  I'll think more as I eat lunch and then dive into the updater functions...

**Comment by davidskalinder** [comment]
Also, I think I can now answer some of @olderwoman's questions based on my testing of the UI:

> multiple coder lines for the same variable (start or end date) and event, which would be generated when the coder returns to edit an event
> ...
> multiple records with start-date or end-date. This can happen when a coder changes their mind.
> ...
> We need to know whether/how coder lines are overwritten or appended if a coder comes back and edits their work. And, of course, the order in which the coding happened.

I can confirm that this is not what's causing this issue, at least not under normal circumstances.  In all the tests I've done, changing a value simply changes the value.  Deleting the text in a text box (such as the start date field) changes the value to an empty string; unchecking a check box (such as the date estimated fields) removes the appropriate row from the database.  So, something else is going on with these bad rows.

We can tell the order of the rows from the auto-incrementing table ID, which will help with cleanup in the rare cases when values differ.

> is a record generated when a coder deletes marked text in the text-marking fields?

In the tests I've done, no: the record that was previously there goes away.  (Indeed, I've just noticed that sometimes some extra ones do too, which I've logged as #89; but that is almost certainly unrelated to this issue.)

So in all cases I can find, the behavior is what we'd expect: to show a coder the most up-to-date information that s/he has entered.  As I mentioned upthread, my guess is that some strange edge case is causing this.  Time for me to read some code!

**Comment by davidskalinder** [comment]
Grunt:

```/* Count of rows with multiple-valued article variables */
SELECT a.variable, count(*)
  FROM coder_article_annotation AS a
  INNER JOIN
    (SELECT coder_id
       , article_id
       , variable
       , COUNT(id) AS n
       FROM coder_article_annotation
       GROUP BY coder_id
       , article_id
       , variable
       HAVING n > 1
     ) AS b
  ON a.coder_id = b.coder_id
  AND a.article_id = b.article_id
  AND a.variable = b.variable
  GROUP BY a.variable;
```
```
+--------------+----------+
| variable     | count(*) |
+--------------+----------+
| no-us-events |        2 |
+--------------+----------+
1 row in set (0.01 sec)
```
```
/* Rows with non-text-capture multi-entry article variables */
SELECT *
  FROM coder_article_annotation AS a
  INNER JOIN
    (SELECT coder_id
       , article_id
       , variable
       , COUNT(id) AS n
       FROM coder_article_annotation
       WHERE RIGHT(variable, 5) != "-text"
       GROUP BY coder_id
       , article_id
       , variable
       HAVING n > 1
     ) AS b
  ON a.coder_id = b.coder_id
  AND a.article_id = b.article_id
  AND a.variable = b.variable;
```
```
+-----+------------+--------------+-------+------+----------+---------------------+----------+------------+--------------+---+
| id  | article_id | variable     | value | text | coder_id | timestamp           | coder_id | article_id | variable     | n |
+-----+------------+--------------+-------+------+----------+---------------------+----------+------------+--------------+---+
| 967 |       3465 | no-us-events | yes   | NULL |        7 | 2020-02-26 12:31:25 |        7 |       3465 | no-us-events | 2 |
| 969 |       3465 | no-us-events | yes   | NULL |        7 | 2020-02-26 12:31:30 |        7 |       3465 | no-us-events | 2 |
+-----+------------+--------------+-------+------+----------+---------------------+----------+------------+--------------+---+
```

So the bug with the checkboxes (which I'm increasingly suspicious is a separate problem) also shows up in the article-level data.  It also turns up several times in the testing DB, and even once in the development deployment.

JS handler bug moves ever upward on my suspect list.  Someday I'll have to stop thinking of queries to debug with and actually follow through on my threat to read the code huh

**Comment by davidskalinder** [comment]
Sheesh, it just doesn't look like the JS is really doing very much here.

Here's the HTML that should create the start date field:

https://github.com/davidskalinder/mpeds-coder/blob/7cdc4d8b836ea1c9c2a3ce14d60d665851ce40cb/templates/event-creator-block.html#L29-L37

Here's the listener:

https://github.com/davidskalinder/mpeds-coder/blob/7cdc4d8b836ea1c9c2a3ce14d60d665851ce40cb/static/ec.js#L105

And here's the handler:

https://github.com/davidskalinder/mpeds-coder/blob/7cdc4d8b836ea1c9c2a3ce14d60d665851ce40cb/static/shared.js#L367-L386

Not much space there for things to go wrong it seems like (modulo some subtle JS malarkey)...

On to the controller function...

**Comment by davidskalinder** [comment]
This will be helpful along with `git blame`:

```
/* Earliest instance of a multi-entry start-date variable */
SELECT MIN(a.timestamp)
  FROM coder_event_creator AS a
  INNER JOIN
    (SELECT coder_id
       , event_id
       , variable
       , COUNT(id) AS n
       FROM coder_event_creator
       WHERE variable = "start-date"
       GROUP BY coder_id
       , event_id
       , variable
       HAVING n > 1
     ) AS b
  ON a.coder_id = b.coder_id
  AND a.event_id = b.event_id
  AND a.variable = b.variable;
```

The earliest instance I can find is in the database for the `blacknews` deployment:

```+---------------------+
| MIN(a.timestamp)    |
+---------------------+
| 2017-05-15 10:59:39 |
+---------------------+
1 row in set (0.02 sec)
```

The earliest one from the campus protest DB is a few months after that.

**Comment by davidskalinder** [comment]
Welp, the controller code isn't too complicated either:

https://github.com/davidskalinder/mpeds-coder/blob/7cdc4d8b836ea1c9c2a3ce14d60d665851ce40cb/mpeds_coder.py#L1418-L1448

(That code as well as most of the JS above was committed on 2017-04-27, so it's consistent with the first occurrence two weeks later.)

The only thing that looks remotely fishy here to me is that there are two different commits for the delete and the append calls; so maybe it's possible that the delete could fail and the next append happens anyway?  Although that wouldn't explain why in most cases the second entry is identical to the first, nor how it could be triggered twice in the same second...

**Comment by davidskalinder** [comment]
Ahh-haaah!:

```
/* Latest instance of a multi-entry start-date variable */
SELECT MAX(a.timestamp)
  FROM coder_event_creator AS a
  INNER JOIN
    (SELECT coder_id
       , event_id
       , variable
       , COUNT(id) AS n
       FROM coder_event_creator
       WHERE variable = "start-date"
       GROUP BY coder_id
       , event_id
       , variable
       HAVING n > 1
     ) AS b
  ON a.coder_id = b.coder_id
  AND a.event_id = b.event_id
  AND a.variable = b.variable;
```
```
+---------------------+
| MAX(a.timestamp)    |
+---------------------+
| 2020-05-31 22:41:34 |
+---------------------+
1 row in set (0.03 sec)
```
```
$ more /var/log/apache2/error.log.1
...
[Sun May 31 22:41:34.153733 2020] [wsgi:error] [pid 25289] /home/www/.conda/envs/mpeds/lib/python2.7/site-packages/sqlalchemy/orm/persistence.py:886: SAWarnin
g: DELETE statement on table 'coder_event_creator' expected to delete 1 row(s); 0 were matched.  Please set confirm_deleted_rows=False within the mapper confi
guration to prevent this warning.
[Sun May 31 22:41:34.153783 2020] [wsgi:error] [pid 25289]   (table.description, expected, rows_matched)
```

So, yes, that delete statement failing caused the problem (at least once says the mathematician lol).  Though I don't think it explains the other things that weren't explained in my last comment.

Alas, I fear I will need to return to the SQLAlchemy documentation yet again...

**Comment by davidskalinder** [comment]
Annoyingly, there are no log entries that correspond to the double-entries for checkbox fields.  I'm increasingly confident that that's a different bug, so I'm going to open a new issue for it.

**Comment by davidskalinder** [comment]
After a quick look at https://docs.sqlalchemy.org/en/13/orm/session_basics.html#what-does-the-session-do, I think the problem here is that the delete query works for the SQLAlchemy object but then fails at the database level.  Unfortunately there's nothing in the MySQL error logs within hours of the error above.  Further research is required.

**Comment by olderwoman** [comment]
Thanks for working on this. I’ve been following the saga.

**Comment by davidskalinder** [comment]
Well...  Based on a pretty careful reading of the SQLAlchemy docs, it looks like those delete and add calls should be part of the same transaction rather than being committed separately.  However:

1. I can't find a straight answer on whether SQLAlchemy automatically rolls back a commit when any part of it fails or not.  I guess we could put it in a try-catch block?
  1. This would also let us put some debugging code in the catch section.  Although I'm not really sure how best to log errors so I'd have to read the Flask docs (carefully, probably) to figure that out.
1. Even with the two commits, I can't think of any logic that would cause what we're seeing.
  1. I don't know what could cause a lack of a matching MySQL record after the object has been deleted but not flushed yet.
  1. Even if the MySQL delete does fail, I don't know why the extra record would usually be identical to the first.
    1. Okay, the value to insert would be whatever's shown in the UI; but why would `add()` change the existing record?
  1. I don't know why the extra record sometimes isn't identical but still has a timestamp within a second of the others.

A single commit *might* fix all of this by ensuring that either both operations happen or neither (although if they both fail we'd have to think through what to show the user).  But it makes me nervous not being able to test it since we don't know what causes the half-failure in the first place...

This is a puzzle.  I think the good news is that we can almost certainly rely on the largest ID being correct.  Although hmm we had better check that that's the one the user sees when s/he opens the form huh.

**Comment by davidskalinder** [comment]
Confirmed that when logged on as the user whose records are shown in https://github.com/davidskalinder/mpeds-coder/issues/85#issuecomment-637017690, I see the values with the `highest coder_event_creator.id`.  So even in that weirdest case, what's in the database is what the coder thinks is in the database.

I think this lowers the priority of this issue somewhat, since that means we know where to find all the good data and nobody is seeing any bad data until we export it.  However it would be nice to get a fix before too long since it's needed to properly finish #83.

**Comment by davidskalinder** [comment]
Turns out that the log warning from https://github.com/davidskalinder/mpeds-coder/issues/85#issuecomment-637131365 is only sometimes in the apache logs -- more often missing than not based on eyeballing the last handful of occurrences.

**Comment by davidskalinder** [comment]
Okay, so I have a hypothesis about something that *could* be causing all of this.  The short version: a brief internet outage makes two requests from a coder queue up and then hit the server at almost the same time and get in a race condition.

The longer version: A coder enters the start date field and enters a date.  While she is typing, her internet connection cuts.  She leaves the field, triggering the blur event, so an AJAX request is created to call the controller function, but, like good AJAX, it stays on the client side until the server is available.  Meantime, she enters the field, does nothing, and leaves it again, triggering another AJAX request.  Then the internet comes back on and both requests are sent to the server.  If they can run in parallel, then both delete calls could be made and committed before both inserts are made, resulting in two identical database records with identical or nearly identical timestamps.

However, none of that would trigger the warning about no matching MySQL records.  What could trigger that is if there had been a value in that field that the coder had entered previously.  In that case, the first delete commit would erase the record, whereas the second delete query would delete the record from SQLAlchemy's session and then be surprised to find that the record is gone when it tries to commit, raising the warning.  After that both records are added as before.

In rare cases, the coder could change the value in the text input box during the outage, resulting in several different values with the same timestamp, as in https://github.com/davidskalinder/mpeds-coder/issues/85#issuecomment-637017690 (which was a coder who might have been working from India at the time, so latency issues are plausible).

It's a little odd that this would happen only for the start and end date fields, but that could be because they're the most frequently-used text fields, or perhaps because the `datetimepicker` gizmo slows down the connection or something.  And I don't think it would explain the slightly different timestamps in #90 (but hey, that's a different issue!).

So maaybe I could test this hypothesis by loading a page, cutting my wifi, leaving the field twice, and then turning the connection back on?  Though it's possible that what's actually causing it is some other kind of connection that's harder to control...

**Comment by davidskalinder** [comment]
> So maaybe I could test this hypothesis by loading a page, cutting my wifi, leaving the field twice, and then turning the connection back on? Though it's possible that what's actually causing it is some other kind of connection that's harder to control...

Yeah so that didn't work.  Maybe AJAX is smart enough to handle that kind of outage more elegantly?  So even if my hypothesis is right, it could be hard to figure out how to test this.

Anyway, I think fixing the two commits removes even the potential for a race condition, so that's probably what we should do.  @alexhanna, do you know of a good way to send debug output to a log so that we can do that if the exception is caught?

**Comment by olderwoman** [comment]
FYI John Lemke seems not to have been getting the last dozen or more of these updates as you have been working through this.

**Comment by davidskalinder** [comment]
> FYI John Lemke seems not to have been getting the last dozen or more of these updates as you have been working through this.

Hah, a lot more than a dozen!  Anyway mentioning @johnklemke here should send him a notification for new posts, and he can review the whole detective story above when convenient...

**Comment by johnklemke** [comment]
Yeah, looked it over -- and yikes!

**Comment by davidskalinder** [comment]
Well, for the life of me I can't find any documentation guaranteeing the order that SQLAlchemy will execute statements when it does a `commit()`.  I guess the only way to proceed is to test it and hope it never changes.  (If the order isn't right, then maybe `flush()`ing after each step will force the correct order?)

**Comment by davidskalinder** [comment]
As discussed in https://github.com/davidskalinder/mpeds-coder/issues/87#issuecomment-638444545 ff., a workaround for this is to simply throw away all but the last value of these fields before analyzing the data.  (It looks like this is always the entry that's visible to the coders, and if so will always be the best entry available.)

So I'm going to take this issue out of the epics for the export issues since that workaround should decouple them.

**Comment by davidskalinder** [comment]
>     1. I can't find a straight answer on whether SQLAlchemy automatically rolls back a commit when any part of it fails or not.  I guess we could put it in a try-catch block?
> 
>     2. This would also let us put some debugging code in the catch section.  Although I'm not really sure how best to log errors so I'd have to read the Flask docs (carefully, probably) to figure that out.

So thinking more about this, I'm not even sure that a, ahem, try-*except* block would do us any good here.

First, the message that gets into the error log is a warning, not an exception.  There [might](https://stackoverflow.com/questions/5644836/in-python-how-does-one-catch-warnings-as-if-they-were-exceptions) be a way to catch the warning like an exception...

But a more serious problem is that very often there's not even a warning -- if there's no preexisting record to delete, then SQLA just happily inserts two records.  I'm not even sure there's a way to check for this at all, since peeking to see whether two records are being inserted would just create another race condition.

> A single commit _might_ fix all of this by ensuring that either both operations happen or neither (although if they both fail we'd have to think through what to show the user). But it makes me nervous not being able to test it since we don't know what causes the half-failure in the first place...

So I guess this is the only game in town?  I still don't know any way of testing it aside from simply deploying and then running [this query](https://github.com/davidskalinder/mpeds-coder/issues/85#issuecomment-637131365) after like a week and seeing if there are any new occurrences?  (After testing to make sure that it doesn't break anything under normal circumstances, of course...)

**Comment by davidskalinder** [comment]
Well, git commit 2db3744 changes the `changeCode()` function to use a single database `commit()` instead of one each for the deletes and the insert.  However, I'm nervous about rashly deploying this given how hard it is to test.  I don't have more time to work on this this week, so maybe it's best to discuss during the meeting on Monday?

In the meantime: @alexhanna, do you recall whether there's some sensible reason why there are two commits in that function?

https://github.com/davidskalinder/mpeds-coder/blob/7cdc4d8b836ea1c9c2a3ce14d60d665851ce40cb/mpeds_coder.py#L1418-L1448

**Comment by davidskalinder** [comment]
Incidentally, there are a bunch of other similar functions that likewise have two `commit()`s.  If this is indeed the problem, then we probably want to fix those too -- though again, without knowing how to test, I think it's better to go cautiously and only change one thing at a time...

**Comment by davidskalinder** [comment]
> ... I'm nervous about rashly deploying this given how hard it is to test. I don't have more time to work on this this week, so maybe it's best to discuss during the meeting on Monday?

Whoops, I forgot that this was the thing to discuss on Monday.  But maybe it's just as well considering that this now seems fairly low priority and given the complications with #89.  So I should try to re-raise this next Monday.  I'll put the question flag on it to hopefully remind me more better.

**Comment by davidskalinder** [comment]
And now that I forgot it again, I added it to next week's meeting's agenda.  So I'll remove the question label.

**Comment by davidskalinder** [comment]
Okay, so the consensus was to go ahead and deploy this and monitor how it goes.  I'm going to at least put it in testing first and have a play with it to make sure nothing unexpected happens immediately, but then I'll make it live and watch it over the next couple weeks...

**Comment by davidskalinder** [comment]
All right, this is deployed in all three installations.  I've tested date entry, deletion, and re-entry in the dev and testing deployment, and refreshed the production deployment to confirm that nothing is obviously broken.

These I think are the outstanding things to do for this issue, in order:

1. Stay paranoid over the next few days in case this breaks anything obvious
1. Check back in a few weeks to see whether this has made the double-entry text bug go away
1. PR 2db3744 upstream
1. Fix all the other instances of double-commits (which might also fix #87)
1. Repeat

**Comment by davidskalinder** [comment]
Well, uh...

```
mysql> /* Rows with non-text-capture multi-entry variables since change */
mysql> SELECT *
    ->   FROM coder_event_creator AS a
    ->   INNER JOIN
    ->     (SELECT coder_id
    ->        , event_id
    ->        , variable
    ->        , COUNT(id) AS n
    ->        FROM coder_event_creator
    ->        WHERE RIGHT(variable, 5) != "-text"
    ->        GROUP BY coder_id
    ->        , event_id
    ->        , variable
    ->        HAVING n > 1
    ->      ) AS b
    ->   ON a.coder_id = b.coder_id
    ->   AND a.event_id = b.event_id
    ->   AND a.variable = b.variable
    ->   WHERE a.timestamp > "2020-06-23 15:00"
    ->   ;
+--------+------------+----------+------------+------------+------+----------+---------------------+----------+----------+------------+---+
| id     | article_id | event_id | variable   | value      | text | coder_id | timestamp           | coder_id | event_id | variable   | n |
+--------+------------+----------+------------+------------+------+----------+---------------------+----------+----------+------------+---+
| 101759 |       1356 |     6645 | start-date | 2006-12-17 | NULL |        5 | 2020-06-23 18:20:53 |        5 |     6645 | start-date | 2 |
| 101760 |       1356 |     6645 | start-date | 2006-12-17 | NULL |        5 | 2020-06-23 18:20:53 |        5 |     6645 | start-date | 2 |
| 101786 |       1356 |     6648 | start-date | 2006-12-17 | NULL |        5 | 2020-06-23 18:23:27 |        5 |     6648 | start-date | 2 |
| 101787 |       1356 |     6648 | start-date | 2006-12-17 | NULL |        5 | 2020-06-23 18:23:27 |        5 |     6648 | start-date | 2 |
| 104988 |       4175 |     6860 | end-date   | 2007-03-25 | NULL |        5 | 2020-06-30 17:25:17 |        5 |     6860 | end-date   | 2 |
| 104989 |       4175 |     6860 | end-date   | 2007-03-25 | NULL |        5 | 2020-06-30 17:25:17 |        5 |     6860 | end-date   | 2 |
| 105001 |       4199 |     6861 | start-date | 2007-04-24 | NULL |        5 | 2020-06-30 17:35:06 |        5 |     6861 | start-date | 2 |
| 105002 |       4199 |     6861 | start-date | 2007-04-24 | NULL |        5 | 2020-06-30 17:35:06 |        5 |     6861 | start-date | 2 |
+--------+------------+----------+------------+------------+------+----------+---------------------+----------+----------+------------+---+
8 rows in set (0.16 sec)
```

So changing to a single commit has not fixed this problem.

I think I'm officially stumped.  My inclination is to revert the change to a single commit (for fear it might break something new, even though a single commit does smell better to me?) and then I guess to forget the whole thing since it can be easily worked around?

Super strange, man.

**Comment by davidskalinder** [comment]
Also, incidentally:

```
mysql> /* Rows with non-text-capture non-date-related multi-entry variables */
mysql> SELECT *
    ->   FROM coder_event_creator AS a
    ->   INNER JOIN
    ->     (SELECT coder_id
    ->        , event_id
    ->        , variable
    ->        , COUNT(id) AS n
    ->        FROM coder_event_creator
    ->        WHERE RIGHT(variable, 5) != "-text"
    ->        AND LEFT(variable, 6) != "start-"
    ->        AND LEFT(variable, 4) != "end-"
    ->        GROUP BY coder_id
    ->        , event_id
    ->        , variable
    ->        HAVING n > 1
    ->      ) AS b
    ->   ON a.coder_id = b.coder_id
    ->   AND a.event_id = b.event_id
    ->   AND a.variable = b.variable
    ->   ;
+-------+------------+----------+----------+-------+------+----------+---------------------+----------+----------+----------+---+
| id    | article_id | event_id | variable | value | text | coder_id | timestamp           | coder_id | event_id | variable | n |
+-------+------------+----------+----------+-------+------+----------+---------------------+----------+----------+----------+---+
| 95662 |       6205 |     6247 | state    | OR    | NULL |       17 | 2020-06-15 04:45:47 |       17 |     6247 | state    | 2 |
| 95664 |       6205 |     6247 | state    | OH    | NULL |       17 | 2020-06-15 04:45:47 |       17 |     6247 | state    | 2 |
+-------+------------+----------+----------+-------+------+----------+---------------------+----------+----------+----------+---+
2 rows in set (0.15 sec)
```

So this has now happened once for a dropdown (which is handled the same way as a radio), as well as for text as described in this issue and for checkboxes as described in #90.  So we should probably expect that this can happen for any variable, really (if rarely).

**Comment by davidskalinder** [comment]
Also, just to note this here: for reasons I don't entirely understand at the moment (probably just the way the application parses the list of values), the UI does show the coder the _last_ entry they've put into the field, which is actually the best possible behavior given that there's actually two entries in there.  So in all cases we should treat the entry with the highest ID as the coder's intent and disregard the others.

**Comment by davidskalinder** [comment]
> My inclination is to revert the change back to two commits (for fear it might break something new, even though a single commit does smell better to me?) and then I guess to forget the whole thing since it can be easily worked around?

Outcome of team discussion: yes, do this.  @johnklemke should be able to handle the duplicates in pass 2 without too much trouble, and nothing of value should be lost.

So I'll revert (well, probably just change back, technically) the changes I made; and then I need to pay attention to what's in which branches and whether anything needs to be PR'd to @alexhanna et al.

**Comment by davidskalinder** [comment]
> So I'll revert (well, probably just change back, technically) the changes

Hahaha disregard that: I did just revert it in the end, in 54eeeb40.  This has been deployed everywhere without breaking anything obvious.

I thiink that since this revert means that nothing was ultimately changed for this issue, there's nothing to revert.  Worst-case, both the change and the revert might accidentally make it upstream along with some other revision?...  But @alexhanna, if you'd prefer that I PR this non-change now to better ensure that the commit histories stay in sync, let me know.

I guess I'll close this issue now, even though the "resolution" has ultimately been to give up with no resolution?  We should reopen if we find more evidence of what's actually going on here, or if the PR is in fact necessary and becomes complicated...

**Comment by davidskalinder** [comment]
While testing #96, I managed to sort-of-reliably reproduce this for the first time, by entering very long (like upwards of 2000-character-) strings into a description field many times (either blanking the field between pastes, or changing long lengths).  This again makes this seem like a latency issue; I'm going to spend a short time debugging further now that it's all set up, before giving up again lol.

**Comment by davidskalinder** [comment]
Here is fun:

When I enter a 4000-character message into the event description field *and leave the field by pressing alt-tab to get to another window*, a double entry happens *almost* every time.  It isn't quite every time; sometimes the double entry only goes in the second time in a row (with no intervening blur events), sometimes it takes longer.  And once a double entry is in the DB, all subsequent alt-tab blurs in a row with long payloads seem to also create double entries.  (Zero-length payloads return it to a single entry.)  I don't think I've ever managed to get it back to a single entry by only doing alt-tab blurs after it's doubled (though this doesn't mean it's impossible).

However, blurring the field any other way -- via a mouse click to the article text, via a mouse click to another application, via tabbing to the next field, via hitting ctrl-tab, or even via hitting alt-shift-tab -- results in a single entry (even if there was a double entry before)!  I don't think I've ever managed to produce a double entry by doing anything other than an alt-tab blur.

And btw, commenting out the first commit in that function doesn't seem to make a lick of difference.

So this makes it look like the bug is some combination of performance (slowed down by lots of text, or maybe something else) interacting with using alt-tab to blur a field.  I'm looking at you, JavaScript.

As promised, I'm going to stop debugging this now.  But this issue's craziness continues.

**Comment by davidskalinder** [comment]
Oh, and also, using `return jsonify(result=str(len(a)))` to see how many entries are already in the DB yields no useful information: it's always 1 until two entries show up in the DB, then it's 2 (no 0s, so no evidence that it's a race condition).

**Comment by davidskalinder** [comment]
One more note on this: when the bug happens (only), it is the case that there are two XHRs that are 100% identical (except for a page number that seems to increment with every additional XHR from a page no matter what).

So the bug certainly does involve an extra request being mistakenly sent from the client to the server, though I don't know why that happens or why it causes the behavior that it does.

I suppose it's possible that even with a single commit there's still something going wrong with the DB transactions that allows the statements to run in parallel instead of in series?...  But if so, I don't understand why.
